### ISSUE TITLE
Doesn't render until page refresh

### ISSUE THREAD

**Issue by cmtonkinson** [kasperpeulen/discourse-mathjax#31]
I have an up-to-date Discourse (and MathJax plugin) installation.

MathJax does not render when navigating to a Topic via link (such as clicking on a Topic from the index page) but _does_ render when the Topic is shown on a page load/refresh (such as navigating to a Topic directly via URI or after a page refresh).

**Comment by cmtonkinson** [comment]
Seems to be related to [this SO question](http://stackoverflow.com/q/20832157/77047) (even though it's specific to MathJax in an Angular app).

**Comment by mcmcclur** [comment]
I have the same problem on my tiny, private discourse:
http://discourse.marksmath.org/

Strangely, if I navigate to a discussion page directly, the MathJax works on first load. The only time the it doesn't work is when I navigate to a page from within Discourse.

**Comment by Blaisorblade** [comment]
TL;DR. It *seems* this happens on Discourse 1.5 because the needed plugin APIs were rewritten and are incompatible. Details also explain why the issue appears only when the page is loaded through AJAX (say, by clicking a link within Discourse), not directly.
Alas, I'm not a JavaScript programmer, I just have a theoretical clue how this works, so consider this an educated guess.

I get the same too (on a private discourse), it's also documented here:
https://meta.discourse.org/t/mathjax-plugin-supports-math-notation-using-latex/12826/69?u=blaisorblade

That happens (for me) using Discourse 1.5.1. Can you confirm you observe the bug on Discourse on 1.5? Maybe @cmtonkinson was on a prerelease of 1.5?

> Strangely, if I navigate to a discussion page directly, the MathJax works on first load. The only time the it doesn't work is when I navigate to a page from within Discourse.

Well, @cmtonkinson's link is helpful even there. In particular, the explanation on this answer makes lots of sense: http://stackoverflow.com/a/30386723/53974
> Because MathJax loads asynchronously (check out the MathJax Doc) , running MathJax.Hub.Queue(["Typeset",MathJax.Hub]); will wait until the page reloads to re-typeset the page. Running the same command again causes the typeset to occur immediately.

What I get is: if some JavaScript updates the webpage content (say, Discourse through AJAX), it should tell MathJax to reprocess the updated content.

Of course, that's always been needed, and variants of `MathJax.Hub.Queue(["Typeset",MathJax.Hub])` *are* called by this plugin — in particular by callbacks registered for some events:

https://github.com/kasperpeulen/discourse-mathjax/blob/master/assets/javascripts/initializers/mathjax.js.es6#L7

So I guess that now, for some reason, one needs to listen to different events.
Since the code that updates the page to render a topic was rewritten in 1.5, this breakage is unsurprising. The blog post announcing the new API shows how to hook a plugin to onPageChange:

https://meta.discourse.org/t/a-new-versioned-api-for-client-side-plugins/40051

https://eviltrout.com/2016/02/25/fixing-android-performance.html http://blog.discourse.org/2016/04/discourse-1-5-released/

**Comment by cmtonkinson** [comment]
@Blaisorblade FYI I'm now on 1.6.0-beta2 (specifically commit 706ea28) but this behavior has been consistent since at _least_ 1.5.

**Comment by Blaisorblade** [comment]
@cmtonkinson I prototyped a fix, but be careful as I'm no JavaScript or Discourse developer — though I *have* taken a look at some docs and done some testing, and the new code looks cleaner.
But don't try this in production yet.
https://github.com/Blaisorblade/discourse-mathjax/tree/topic/discourse-1.5-v3

Scenarios I tested, grouped by distinct code paths being exercised:
1. get to a topic from the topic list
2. get to a topic by refreshing the page
3. edit an existing post, or reply to a post, and check that MathJax works in the preview

Untested:
- compatibility with older Discourse releases, though I followed instructions so there's a good chance it *should* work.

Possible regressions:
- the old code has a snippet ensuring that "if the caret is on the last line ensure preview scrolled to bottom". I left that out (for now) of the new version, because I'm not sure of the proper way to port it (and I don't get the point anyway, that seems the job of Discourse itself so it seems a hack; but I haven't studied the issue).

Bugs I had at some point and then *fixed*:
- rendered output appearing twice
- rendered output and rest of the text being mixed together

Both bugs were fixed by following the documented API more carefully, but please watch out for them (or other ones) in testing.

**Comment by Blaisorblade** [comment]
For the record: I'm now happily using my PR (#32) in production.
